src/main/java/io/confluent/connect/jdbc/source/TimestampIncrementingCriteria.java

```diff
@@ -218,17 +220,26 @@ public TimestampIncrementingOffset extractValues(
    *
    * @param schema the record's schema; never null
    * @param record the record's struct; never null
+   * @param timestampGranularity defines the configured granularity of the timestamp field
    * @return the timestamp for this row; may not be null
    */
   protected Timestamp extractOffsetTimestamp(
       Schema schema,
-      Struct record
+      Struct record,
+      JdbcSourceConnectorConfig.TimestampGranularity timestampGranularity
   ) {
     caseAdjustedTimestampColumns.computeIfAbsent(schema, this::findCaseSensitiveTimestampColumns);
     for (String timestampColumn : caseAdjustedTimestampColumns.get(schema)) {
-      Timestamp ts = (Timestamp) record.get(timestampColumn);
-      if (ts != null) {
-        return ts;
+      try {
+        Timestamp ts = timestampGranularity.toTimestamp.apply(record.get(timestampColumn));
+        if (ts != null) {
+          return ts;
+        }
+      } catch (NumberFormatException  e) {
+        throw new ConnectException(
```

Comment:
Can this be caught and handled only in the enum `toTimestamp` method that would throw the exception?